File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <map>
using namespace std;

vector<string> solution(vector<string> p, vector<string> c) {
    vector<string> answer;
    map<string, int> m;
    
    for (int i = 0; i < p.size(); i++)
        m[p[i]] = i;

    for (int i = 0; i < c.size(); i++) {
        int called = m[c[i]];
        m[p[called - 1]]++;
        m[p[called]]--;
        swap(p[called - 1], p[called]);
    }
    
    return p;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        String[] answer = {};
        Map<String, Integer> map = new HashMap<>();
        List<String> arr = new ArrayList<>();
        
        for (int i = 0; i < players.length; i++) {
            map.put(players[i], i);
            arr.add(players[i]);
        }
     
        for (String call : callings) {
            int idx = map.get(call);
            String overName = arr.get(idx - 1);
            
            arr.remove(idx);
            arr.add(idx - 1, call);

            map.put(call, idx - 1);
            map.put(overName, idx);
        }
        
        return arr.stream().toArray(String[]::new);
    }
}